### ISSUE TITLE
could handle 300-level responses better

### ISSUE THREAD

**Issue by davepacheco** [oxidecomputer/progenitor#451]
I have an API endpoint that returns a 303 redirect on success with an empty body.  The response contains a header that I want to be able to get at.  I'm not sure this is intended to be supported with Progenitor.  However, the current failure mode is:

```
    Unexpected Response: Response { url: Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Ipv4(127.0.0.1)), port: Some(12220), path: "//login/test-suite-silo/local", query: None, fragment: None }, status: 303, headers: {"location": "/", "set-cookie": "session=cdb10d5430e901b00050ff41480ad0afdc6869ee; Path=/; HttpOnly; SameSite=Lax; Max-Age=28800", "x-request-id": "6a543c5a-bc48-43ac-8cb2-b0352d08b550", "content-length": "0", "date": "Tue, 02 May 2023 00:00:53 GMT"} }
```

Here's the spec for this endpoint:

```
    "/login/{silo_name}/local": {
      "post": {
        "tags": [
          "login"
        ],
        "summary": "Authenticate a user via username and password",
        "operationId": "login_local",
        "parameters": [
          {
            "in": "path",
            "name": "silo_name",
            "required": true,
            "schema": {
              "$ref": "#/components/schemas/Name"
            }
          }
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/UsernamePasswordCredentials"
              }
            }
          },
          "required": true
        },
        "responses": {
          "303": {
            "description": "redirect (see other)",
            "headers": {
              "location": {
                "description": "HTTP \"Location\" header",
                "style": "simple",
                "required": true,
                "schema": {
                  "type": "string"
                }
              }
            }
          },
          "4XX": {
            "$ref": "#/components/responses/Error"
          },
          "5XX": {
            "$ref": "#/components/responses/Error"
          }
        }
      }
    },
```

which shows that it has a 303 response.

It might be fair to say that 300-level responses aren't supported, especially since reqwest follows 300s by default, so in principle Progenitor wouldn't usually see them.  In this case, I configured my reqwest client _not_ to follow the redirect so that I could get the header I want back.  If that's not supported though then I think we should at least produce a clearer error (e.g., `300-level response codes are not supported`).  We may even want to fail to generate a client for a spec with a 300-level response?  Or emit a warning and skip that endpoint altogether?

@ahl noticed we don't handle 300-level responses here:
https://github.com/oxidecomputer/oxide.rs/blob/d007f734e8f379d8a174d45c18aabd67ec24eef5/sdk/src/generated_sdk.rs#L20053-L20062

Gory details (which I hope aren't necessary, but if you want to repro this): https://github.com/davepacheco/oxide-client-test

**Comment by jclulow** [comment]
I suspect various APIs probably also use this stuff, so if it's able to be expressed well in an OpenAPI document we probably do want to support it.

**Comment by ahl** [comment]
I often use this repo to get a sense of what's out there in the wild: https://github.com/APIs-guru/openapi-directory I didn't find a single instance of a 3xx response code (though it's not impossible that my search criteria were insufficient). It's not particularly well-expressed in OpenAPI in that we don't know anything about the responses from the URL to which we're redirected. There's a good articulation of the problem here: https://github.com/OAI/OpenAPI-Specification/issues/2512

If we were to support it, a viable path would be as @davepacheco suggests: treat 3xx responses effectively as 2xx responses and don't automatically follow them. 

In this particular example do we want a redirect response? It seems like this might inhibit, say, the CLI from making use of this type of login.

**Comment by davepacheco** [comment]
> In this particular example do we want a redirect response?  It seems like this might inhibit, say, the CLI from making use of this type of login.

Yeah, good questions.  This is complicated.  This issue came up in the context of the password login endpoint, which is (currently) parallel to the SAML login endpoint.  From the server perspective, they look the same.  Only the request body format is different.  There are two related questions: (1) should these be supported by progenitor at all? and (2) should they return a redirect?

On the one hand, these endpoints are really only intended for the browser.  They return one thing, a new session token, and they do so as a cookie (header) on the response.   Non-browser clients are supposed to use the OAuth device auth flow.  (That requires custom client and CLI code, right?  Is _that_ the answer here -- that every supported login mechanism has a small bit of custom client code?  In the future we might also support something else, like service account credentials.  I'm not sure yet how that would look (e.g., do you provide some kind of bearer token as a header with each request or do you use the token to get back a session token?).)

On the other hand, even if we decide that these endpoints are not intended for use outside the browser, it's useful for automated testing if Progenitor can support them.  I could see saying it's not worth it though.

On the question of redirects: in the SAML case, with a browser client, I believe you have to use a redirect because the request is being sent by the browser while on the IdP's site.  We need a redirect to send the browser back to our console.  For username/password login, I think a redirect isn't strictly necessary because we'll be serving the login page from our own site.  We might still want to redirect the user (e.g., from /login to /projects or something), but that can probably be done from the client side as well.  I think I made the password login endpoint use a redirect because on the backend it looks exactly parallel to the SAML case.

So I could see:

1. Have Progenitor fully support 300-level responses without following the redirect.  You'd presumably get back a `ResponseValue` that lets you look at the response but otherwise has no response body.  This would probably still be a kind of awkward because you'd have to have configured the reqwest client to do the right thing first.  (I wonder if the Client constructor could/should take a reqwest client _builder_ so that it could customize it.) 
2. Explicitly don't support APIs with 300-level responses.  This seems reasonable if we think these aren't widely used.  It's a little annoying for the few cases where we _want_ to exercise these endpoints programmatically (mostly in test).  Anyway, if we do this, I imagine we'd want to:
    a. Have Progenitor fail more clearly at runtime when encountering a 300-level response.
    b. Have Progenitor skip generating functions (and types?) for endpoints that produce 300-level responses.
    c. (unrelated to Progenitor): consider removing these endpoints from the spec altogether.  This might be annoying for a couple of existing spec consumers like the console and some of the automated testing (e.g., the authz coverage test).

Thoughts?

**Comment by ahl** [comment]
Agreed with @davepacheco regarding the two viable options. Given that `reqwest` allows for control over following redirects on the `Client` rather than the `RequestBuilder` I'm more inclined towards option number 2: for operations with a 3xx response progenitor can skip generation of related methods (and types).
